File: src/dataHealth/cleanAppDataView.ts
import type { AppData, ScreeningProfile, TrainingSession } from '../models/training-model';
import {
  applyDurationGuard,
  applyHealthDataGuard,
  applyIssueScoreCap,
  applyPerformanceDropGuard,
  applySessionLifecycleGuard,
  applyTodayStatusGuard,
  defaultGuardClock,
  stripLegacyAdviceFromSession,
  type DurationGuardOutcome,
  type HealthDataGuardOutcome,
  type IssueScoreCapOutcome,
  type PerformanceDropOutcome,
  type RuntimeGuardClock,
  type TodayStatusGuardOutcome,
} from './dataHealthRuntimeGuard';

export interface CleanAppDataView {
  raw: AppData;
  appData: AppData;
  durations: Map<string, DurationGuardOutcome>;
  todayStatus: TodayStatusGuardOutcome;
  healthData: HealthDataGuardOutcome;
  issueScoreCap: IssueScoreCapOutcome;
  performanceDrops: PerformanceDropOutcome;
  guardDiagnostics: CleanAppDataViewDiagnostics;
}

export interface CleanAppDataViewDiagnostics {
  lifecycleResidueSessionIds: string[];
  legacyAdviceSessionIds: string[];
  invalidDurationSessionIds: string[];
  cappedIssueScoreKeys: string[];
  staleTodayStatus: boolean;
  staleHealthData: boolean;
  filteredPerformanceDropIds: string[];
}

const buildCleanScreening = (
  screening: ScreeningProfile | undefined,
  issueScoreCap: IssueScoreCapOutcome,
  performanceDrops: PerformanceDropOutcome,
): ScreeningProfile | undefined => {
  if (!screening) return screening;
  const adaptiveState = screening.adaptiveState;
  if (!adaptiveState) return screening;
  return {
    ...screening,
    adaptiveState: {
      ...adaptiveState,
      issueScores: issueScoreCap.cappedScores,
      performanceDrops: performanceDrops.filteredDrops,
    },
  };
};

const buildCleanSession = (session: TrainingSession): {
  session: TrainingSession;
  durationOutcome: DurationGuardOutcome;
  residueChanged: boolean;
  legacyChanged: boolean;
} => {
  const lifecycle = applySessionLifecycleGuard(session);
  const stripped = stripLegacyAdviceFromSession(lifecycle.session);
  const durationOutcome = applyDurationGuard(stripped.session);
  let next = stripped.session;
  if (durationOutcome.derivedDurationMin !== undefined) {
    next = { ...next, durationMin: durationOutcome.derivedDurationMin };
  }
  if (durationOutcome.durationInvalid) {
    next = {
      ...next,
      durationMin: undefined,
    } as TrainingSession & { durationInvalid?: boolean };
    (next as TrainingSession & { durationInvalid?: boolean }).durationInvalid = true;
  }
  return {
    session: next,
    durationOutcome,
    residueChanged: lifecycle.changed,
    legacyChanged: stripped.changed,
  };
};

export const buildCleanAppDataView = (
  rawAppData: AppData,
  clock: RuntimeGuardClock = defaultGuardClock,
): CleanAppDataView => {
  const durations = new Map<string, DurationGuardOutcome>();
  const lifecycleResidueSessionIds: string[] = [];
  const legacyAdviceSessionIds: string[] = [];
  const invalidDurationSessionIds: string[] = [];

  const cleanedHistory = (rawAppData.history || []).map((session) => {
    const built = buildCleanSession(session);
    durations.set(session.id, built.durationOutcome);
    if (built.residueChanged) lifecycleResidueSessionIds.push(session.id);
    if (built.legacyChanged) legacyAdviceSessionIds.push(session.id);
    if (built.durationOutcome.durationInvalid) invalidDurationSessionIds.push(session.id);
    return built.session;
  });

  const cleanedActiveSession = rawAppData.activeSession
    ? buildCleanSession(rawAppData.activeSession).session
    : rawAppData.activeSession;

  const todayStatus = applyTodayStatusGuard(rawAppData, clock);
  const healthData = applyHealthDataGuard(rawAppData, clock);
  const issueScoreCap = applyIssueScoreCap(rawAppData.screeningProfile);
  const performanceDrops = applyPerformanceDropGuard(rawAppData.screeningProfile, cleanedHistory);
  const cleanedScreening = buildCleanScreening(rawAppData.screeningProfile, issueScoreCap, performanceDrops);

  const cleanedAppData: AppData = {
    ...rawAppData,
    history: cleanedHistory,
    activeSession: cleanedActiveSession,
    screeningProfile: cleanedScreening || rawAppData.screeningProfile,
  };

  const diagnostics: CleanAppDataViewDiagnostics = {
    lifecycleResidueSessionIds,
    legacyAdviceSessionIds,
    invalidDurationSessionIds,
    cappedIssueScoreKeys: issueScoreCap.changes.map((entry) => entry.key),
    staleTodayStatus: todayStatus.ignoredForCurrentReadiness,
    staleHealthData: healthData.staleForReadiness,
    filteredPerformanceDropIds: performanceDrops.removed,
  };

  return {
    raw: rawAppData,
    appData: cleanedAppData,
    durations,
    todayStatus,
    healthData,
    issueScoreCap,
    performanceDrops,
    guardDiagnostics: diagnostics,
  };
};

export const cleanAppDataViewIsDirty = (view: CleanAppDataView): boolean => {
  const d = view.guardDiagnostics;
  return (
    d.lifecycleResidueSessionIds.length > 0 ||
    d.legacyAdviceSessionIds.length > 0 ||
    d.invalidDurationSessionIds.length > 0 ||
    d.cappedIssueScoreKeys.length > 0 ||
    d.staleTodayStatus ||
    d.staleHealthData ||
    d.filteredPerformanceDropIds.length > 0
  );
};

File: ios/packages/IronPathDataHealth/Sources/IronPathDataHealth/CleanAppDataView.swift
// CleanAppDataView — iOS-3A Data Health Runtime Foundation V1.
//
// Read-only projection of an `AppData` value. Carries:
//   * the original `raw` AppData (NEVER mutated)
//   * the per-session DurationGuardOutcomes
//   * the AppData-level today / health / issue / drop outcomes
//   * the in-memory "cleaned" projection of history + activeSession +
//     screeningProfile, built by applying the runtime guards as pure
//     value transformations
//   * a `CleanAppDataViewDiagnostics` summary suitable for the
//     iOS-3B repair-orchestrator to consume without re-running guards
//
// The cleaned projection is built as new Swift value-typed structs —
// nothing in `raw` is rewritten, and `raw.canonicalJSONData()` is
// unaffected. Downstream consumers (iOS-4 TrainingDecision, iOS-5
// Focus Mode) read from CleanAppDataView and never touch raw.history
// directly.
//
// Mirrors `retired web reference`.

import Foundation
import IronPathDomain

public struct CleanAppDataViewDiagnostics: Equatable, Sendable {
    public let lifecycleResidueSessionIds: [String]
    public let legacyAdviceSessionIds: [String]
    public let invalidDurationSessionIds: [String]
    public let cappedIssueScoreKeys: [String]
    public let staleTodayStatus: Bool
    public let staleHealthData: Bool
    public let filteredPerformanceDropIds: [String]

    public init(
        lifecycleResidueSessionIds: [String],
        legacyAdviceSessionIds: [String],
        invalidDurationSessionIds: [String],
        cappedIssueScoreKeys: [String],
        staleTodayStatus: Bool,
        staleHealthData: Bool,
        filteredPerformanceDropIds: [String]
    ) {
        self.lifecycleResidueSessionIds = lifecycleResidueSessionIds
        self.legacyAdviceSessionIds = legacyAdviceSessionIds
        self.invalidDurationSessionIds = invalidDurationSessionIds
        self.cappedIssueScoreKeys = cappedIssueScoreKeys
        self.staleTodayStatus = staleTodayStatus
        self.staleHealthData = staleHealthData
        self.filteredPerformanceDropIds = filteredPerformanceDropIds
    }
}

public struct CleanAppDataView: Equatable, Sendable {
    /// The original AppData value. iOS-3A NEVER mutates this — it is
    /// the source-of-truth for round-trip parity and snapshot hash.
    public let raw: AppData

    /// In-memory cleaned history sessions. Each entry has lifecycle
    /// residue cleared, legacy advice stripped, and duration override
    /// applied where the guard fired. Never written back to `raw`.
    public let cleanedHistory: [TrainingSession]

    /// In-memory cleaned active session (same projection as
    /// `cleanedHistory`), nil if `raw.activeSession` is nil.
    public let cleanedActiveSession: TrainingSession?

    /// In-memory cleaned screening profile — `adaptiveState.issueScores`
    /// values are capped and `adaptiveState.performanceDrops` is
    /// filtered. Identical to `raw.screeningProfile` when no changes.
    public let cleanedScreening: ScreeningProfile

    /// Per-session DurationGuardOutcome keyed by `session.id`. Sessions
    /// without an id are not represented here (matches the legacy web schema behaviour
    /// — legacy web schema keys by `session.id` directly).
    public let durations: [String: DurationGuardOutcome]

    public let todayStatus: TodayStatusGuardOutcome
    public let healthData: HealthDataGuardOutcome
    public let issueScoreCap: IssueScoreCapOutcome
    public let performanceDrops: PerformanceDropOutcome

    public let diagnostics: CleanAppDataViewDiagnostics

    public init(
        raw: AppData,
        cleanedHistory: [TrainingSession],
        cleanedActiveSession: TrainingSession?,
        cleanedScreening: ScreeningProfile,
        durations: [String: DurationGuardOutcome],
        todayStatus: TodayStatusGuardOutcome,
        healthData: HealthDataGuardOutcome,
        issueScoreCap: IssueScoreCapOutcome,
        performanceDrops: PerformanceDropOutcome,
        diagnostics: CleanAppDataViewDiagnostics
    ) {
        self.raw = raw
        self.cleanedHistory = cleanedHistory
        self.cleanedActiveSession = cleanedActiveSession
        self.cleanedScreening = cleanedScreening
        self.durations = durations
        self.todayStatus = todayStatus
        self.healthData = healthData
        self.issueScoreCap = issueScoreCap
        self.performanceDrops = performanceDrops
        self.diagnostics = diagnostics
    }

    /// True if at least one of the diagnostic buckets reported a change.
    /// iOS-3B uses this as the gate before invoking auto-repairs.
    public var hasDirtyData: Bool {
        let d = diagnostics
        if !d.lifecycleResidueSessionIds.isEmpty { return true }
        if !d.legacyAdviceSessionIds.isEmpty { return true }
        if !d.invalidDurationSessionIds.isEmpty { return true }
        if !d.cappedIssueScoreKeys.isEmpty { return true }
        if d.staleTodayStatus { return true }
        if d.staleHealthData { return true }
        if !d.filteredPerformanceDropIds.isEmpty { return true }
        return false
    }
}
